File: consistent-hashing-ii/consistent-hashing-ii.py
import bisect
import random


class Solution:
    
    # @param {int} n a positive integer
    # @param {int} k a positive integer
    # @return {Solution} a Solution object
    @classmethod
    def create(cls, n, k):
        # Write your code here
        solution = cls()
        solution.points = []
        solution.point_machine = {}
        solution.n_shards = n
        solution.n_points = k
        return solution
    
    # @param {int} machine_id an integer
    # @return {int[]} a list of shard ids
    def addMachine(self, machine_id):
        # write your code here
        point_list = []
        for i in range(self.n_points):
            while True:
                tmp = random.randint(0, self.n_shards - 1)
                index = bisect.bisect_left(self.points, tmp)
                if index == len(self.points) or self.points[index] != tmp:
                    bisect.insort(self.points, tmp)
                    self.point_machine[tmp] = machine_id
                    point_list.append(tmp)
                    break
        return point_list
    
    # @param {int} hashcode an integer
    # @return {int} a machine id
    def getMachineIdByHashCode(self, hashcode):
        # write your code here
        index = bisect.bisect_left(self.points, hashcode)
        if index == len(self.points):
            index = 0
        return self.point_machine[self.points[index]]

File: consistent-hashing-ii/consistent-hashing-ii.cc
class Solution {
    public:
    // @param n a positive integer
    // @param k a positive integer
    // @return a Solution object
    static Solution create(int n, int k) {
        // Write your code here
        srand(time(nullptr));
        Solution solution = Solution();
        solution.points.clear();
        solution.point_machine.clear();
        solution.n_machines = n;
        solution.n_points = k;
        return solution;
    }
    
    // @param machine_id an integer
    // @return a list of shard ids
    vector<int> addMachine(int machine_id) {
        // Write your code here
        vector<int> point_list;
        for (int i = 0; i < n_points; ++i) {
            while (true) {
                int tmp = rand() % n_machines;
                if (points.find(tmp) == points.end()) {
                    points.insert(tmp);
                    point_machine[tmp] = machine_id;
                    point_list.push_back(tmp);
                    break;
                }
            }
        }
        return point_list;
    }
    
    // @param hashcode an integer
    // @return a machine id
    int getMachineIdByHashCode(int hashcode) {
        // Write your code here
        set<int>::iterator it = points.lower_bound(hashcode);
        if (it == points.end()) {
            return point_machine[*points.lower_bound(0)];
        } else {
            return point_machine[*it];
        }
    }
    
    public:
    int n_machines;
    int n_points;
    set<int> points;
    map<int, int> point_machine;
};
